authentication/authentication.go

```diff
@@ -0,0 +1,145 @@
+package authentication
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+//nolint:gochecknoglobals
+// authenticatorFactories map is used for the providers' self-registration.
+var authenticatorFactories map[string]AuthenticatorFactory
+
+//nolint:gochecknoglobals
+// authenticatorsMtx is used to protect the authenticatorFactories.
+var authenticatorsMtx sync.RWMutex
+
+//nolint:gochecknoinits
+func init() {
+	authenticatorFactories = make(map[string]AuthenticatorFactory)
+}
+
+// onboardNewProvider is used by the providers to register themselves.
+func onboardNewProvider(providerType string, factory AuthenticatorFactory) {
+	authenticatorsMtx.Lock()
+	defer authenticatorsMtx.Unlock()
+
+	authenticatorFactories[providerType] = factory
+}
+
+// AuthenticatorFactory represents each authenticator factory, a function to create a new corresponding
+// authentication provider.
+type AuthenticatorFactory func(config map[string]interface{}, tenant string, registrationRetryCount *prometheus.CounterVec, logger log.Logger) (Provider, error)
+
+// Provider is an interface that should be implemented to onboard a new authentication.
+// provider.
+type Provider interface {
+	Middleware() Middleware
+	Handler() (string, http.Handler)
+}
+
+type tenantHandlers map[string]http.Handler
+
+// AuthenticatorsHandlers manages all middleware and handlers of all authenticators.
+type AuthenticatorsHandlers struct {
```

Comment:
Naming is hard. My only remaining nit with this PR is that this type doesn't mean anything to me and is somewhat confusing.

The methods exposed by this type are `NewTenantAuthenticator`, `AuthenticatorMiddlewares`, and `PatternRoutes`. 
I think that `NewTenantAuthenticator` should be renamed, because it's not a constructor that returns authenticators (that is what the AuthenticatorFactory interface does) it rather configures an authenticator for a tenant. It's also somewhat confusing that a `AuthenticatorsHandlers` shoud know how to create a `NewTenantAuthenticator`, as from its name it seems to know more about handlers instead.

Morally, this type serves the purpose of maintaining a mapping of tenants <-> authenticators and allows us to get metadata about the authenticator for a certain tenant, i.e. middleware.

Maybe we can instead call this a `ProviderRepository`.

`NewTenantAuthenticator` could instead be called `ConfigureProvider`?

Reply:
hmm naming is hard, I would vote to use InitializeProvider instead, that at least used before with OIDC. For the AuthenticatorsHandlers, we could either call ProviderRepository or maybe ProviderManager? What do you think?
@squat @periklis

Reply:
I like `ProviderManager` :+1: